### ISSUE TITLE
Multiline signatures fixer

### ISSUE THREAD

**Issue by greg0ire** [PHP-CS-Fixer/PHP-CS-Fixer#3168]
Hey! This weekend, I wrote [a blog post](https://dev.to/greg0ire/the-case-for-multiline-signatures-dl6) about multiline signatures, and I think it would be great to have a check for this, based on the line length, or @Soullivaneuh proposed based on the number of arguments (or both).

I can work on it unless someone else is very enthusastic about doing it themselves. What do you think?

**Comment by ostrolucky** [comment]
Whoever is going to code this might want to get inspiration from https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Generic/Sniffs/Files/LineLengthSniff.php

**Comment by keradus** [comment]
Sure, go for it !
yet, it will have to allow both approaches, enfore multiline and enforce singleline

ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1980 , especially last comment

**Comment by julienfalque** [comment]
This would be an awesome feature, but I think it should be a fixer dedicated to reduce any line length, not only method signatures. Something like `LineMaxLengthFixer` that is able to chop down long lines when possible (arrays, method signatures, method calls, ...). Maybe the first version would support method signatures only, and other types of lines support would be added later. WDYT?

**Comment by greg0ire** [comment]
@julienfalque great idea, but I think this is contradictory with

>  yet, it will have to allow both approaches, enfore multiline and enforce singleline

I'm all for it though. WDYT @keradus ?

**Comment by Soullivaneuh** [comment]
IMHO, the multi-line signatures or arrays must have an argument option as I said in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1217#issuecomment-338436961

Please see also the eslint example.

BTW, we may add the max_length option but the max-length condition does not apply only for multi-lines signatures or arrays.

**Comment by julienfalque** [comment]
To be honest I don't like the idea of couting the number of arguments because you may have a signature with a few long parameters:
```php
class Foo
{
    public function someVeryLongMethodName(SomeLongClassName $someLongParameterName, AnotherLongClassName $anotherLongParameterName)
    {
    }
}
```
or with a lot of short parameters:
```php
class Foo
{
    public function getFoo(Bar $bar1, Bar $bar2, Bar $bar3, Bar $bar4, Bar $bar5, Bar $bar6)
    {
    }
}
```

In both examples, a fixer that simply counts the number of arguments to make a decision would probably be wrong (ignoring the first one while it should fix it, fixing the second while it should ignore it).

**Comment by greg0ire** [comment]
Here is a draft of how it might look if we accept @julienfalque 's proposal: https://github.com/greg0ire/PHP-CS-Fixer/commit/da28d7a32d918ed732eae03ca0590bddd796bf73

**Comment by Soullivaneuh** [comment]
@juliendufresne You took very extreme sample here.

Both options should be proposed in this case:

- Elements count for consistency (not having lines-broke method on some place, and not elsewhere).
- Max length for such extreme cases.

In this case, it will fill all needs.

@greg0ire BTW, we need the first option for sonata.

Also, I'm not sure "line length fixer" is the best name to have, because a line length is a global constraint, not a way to fix.

How about, based on eslint rule, `ElementNewLineFixer` with such options?

- `element` choices between (`array_element`, `signature_argument`, `call_argument`, anything else?)
- `min_items`: The minimum items needed to break the line. `false` or `null` if you don't want this option.
- `max_length`: The maximum length needed to break the line. `false` or `null` if you don't want this option.
- Also maybe: `inclusive`: If true, both `min_items` and `max_length` must be respected to break the line. But I'm not sure about this one.

**Comment by greg0ire** [comment]
`LengthyLineFixer` could be a better name than `LineLengthFixer`. I don't like `ElementNewLineFixer`, it's meaningless to me.

Also, I fail to see how the number of items should play a role in deciding whether to linebreak or not. Not sure why eslint decided to implement that.

**Comment by Soullivaneuh** [comment]
> I don't like ElementNewLineFixer, it's meaningless to me.

Length based fixer name is meaningless if we propose both option.

> Also, I fail to see how the number of items should play a role in deciding whether to linebreak or not.

As some don't see why method should be line broken at all. AFAIK, php-cs-fixer is here to provide fixer that you can use (or not) and configure as you wish and do/should not decide how the code should be presented.

> Not sure why eslint decided to implement that.

Certainly by feature request, has it's done here. Feel free to make a PR to eslint. :trollface:

**Comment by ostrolucky** [comment]
Hey everybody, this fixer is available at https://github.com/Symplify/CodingStandard#arguments-should-be-on-the-samestandalone-line-to-fit-line-length (Symplify\CodingStandard\Fixer\LineLength\BreakMethodArgumentsFixer)

**Comment by nicolas-grekas** [comment]
On Symfony, we only want single line signatures, whatever the line length.
It'd save us some annoying review comments if php-cs-fixer could enforce it for us.
:+1: for what it's worth :)

**Comment by keradus** [comment]
Can you raise a PR, @nicolas-grekas ?

**Comment by nicolas-grekas** [comment]
Nope sorry I need help also.

**Comment by julienfalque** [comment]
@nicolas-grekas Since version 2.12, the `method_argument_space` rule has a `on_multiline` option that can be set to `ensure_single_line`.

**Comment by keradus** [comment]
@julienfalque , that means this issue is outdated and we can close it already ?

**Comment by nicolas-grekas** [comment]
@julienfalque thanks for the hint. I tried it on the symfony code base, and it also changes function calls, no only signatures. Only signatures should be forced on a single line. Is there any way to achieve this?

**Comment by julienfalque** [comment]
@nicolas-grekas I don't think this is possible at the moment. Please submit a new feature request.

**Comment by k00ni** [comment]
Any update on this?

**Comment by Wirone** [comment]
Update 2023: there's [`Symplify\CodingStandard\Fixer\LineLength\LineLengthFixer`](https://github.com/symplify/coding-standard/blob/main/docs/rules_overview.md#linelengthfixer) as an alternative solution.

**Comment by ruudk** [comment]
@Wirone Thanks for linking these. I just tried setting these up, but they cannot be used for PHP-CS-Fixer because: 
```
In FixerFactory.php line 129:
                                                                               
  Fixer named "Symplify\CodingStandard\Fixer\AbstractSymplifyFixer" has invalid name.                                                                     
```
```php

$blockfinder = new BlockFinder();
$arrayAnalyzer = new ArrayAnalyzer(new TokenSkipper($blockfinder));
$arrayItemNewliner = new ArrayItemNewliner($arrayAnalyzer, new WhitespacesFixerConfig('    ', "\n"));
$arrayBlockInfoFinder = new ArrayBlockInfoFinder($blockfinder);

return (new Config())
    ->registerCustomFixers([
        new ArrayListItemNewlineFixer($arrayItemNewliner, $arrayAnalyzer, $arrayBlockInfoFinder),
    ])
```

**Comment by Wirone** [comment]
Thanks @ruudk for the feedback! Yeah, I also find ECS approach better, it allows navigating from the config to the fixer and is more explicit. Anyway, Fixer expects that custom fixers have `vendor/intent` naming convention, while `AbstractSymplifyFixer::getName()` returns FQCN.

**Comment by Wirone** [comment]
I'll keep this open, but I believe we should split it to multiple issues or at least description should be refreshed, so DoD is clear. As far as I understand there are multiple concerns here:

- ability to enforce multiline signature: already covered partially by `method_argument_space`, but it only can ensure fully multiline if there's already at least one new line within `()`
- ability to enforce single line in signature: already possible with `on_multiline => ensure_single_line` in `method_argument_space`
- both approaches mentioned above would require separate strategies for signatures and calls, so one can choose how each of them is formatted (currently it's "all or nothing" where applicable), as it seems there's such [requirement](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3168#issuecomment-418017973)
- [generic solution was suggested](#issuecomment-338702027) and it was partially covered in #6930, previously in #6387 (only arrays). I believe it could be good to have generic fixer for this (signatures, calls, arrays) and [this suggestion](#issuecomment-338914134) looks fine as a starting point for the discussion

But since this was created 6 years ago and discussion ended 5 years ago, I kindly ask @greg0ire @soullivaneuh @ostrolucky @nicolas-grekas @julienfalque to provide an update from your side 🙂. I would like to clean this up.

**Comment by greg0ire** [comment]
From my side, I still stand by everything I wrote in my blog post, in fact, I think it is even more relevant today because of the type hinting improvements (union types, intersection types, etc) but I no longer have the motivation to work on implementing it :stuck_out_tongue: 



I don't know how it went for Symfony, in particular I don't know if that policy of having everything on one line is also enforced when constructor property promotion is involved.

**Comment by Wirone** [comment]
FYI: found a nice userland alternative: [`PedroTroller/line_break_between_method_arguments`](https://github.com/PedroTroller/PhpCSFixer-Custom-Fixers?tab=readme-ov-file#pedrotrollerline_break_between_method_arguments). Can be used both for usage and as an inspiration for design/implementation.

**Comment by jlecordier** [comment]
> To be honest I don't like the idea of couting the number of arguments because you may have a signature with a few long parameters:
> 
> ```
> class Foo
> {
>     public function someVeryLongMethodName(SomeLongClassName $someLongParameterName, AnotherLongClassName $anotherLongParameterName)
>     {
>     }
> }
> ```
> 
> or with a lot of short parameters:
> 
> ```
> class Foo
> {
>     public function getFoo(Bar $bar1, Bar $bar2, Bar $bar3, Bar $bar4, Bar $bar5, Bar $bar6)
>     {
>     }
> }
> ```
> 
> In both examples, a fixer that simply counts the number of arguments to make a decision would probably be wrong (ignoring the first one while it should fix it, fixing the second while it should ignore it).

I personally don't care about the character length of the signature, at least 3 parameters on a single line is too much cognitive load for my weird brain, whether the names are long or not

**Comment by soullivaneuh** [comment]
> - ability to enforce multiline signature: already covered partially by `method_argument_space`, but it only can ensure fully multiline if there's already at least one new line within ()
> - ability to enforce single line in signature: already possible with `on_multiline => ensure_single_line` in `method_argument_space`

Indeed, which is quite not consistent regarding the `'on_multiline' => 'ensure_fully_multiline'` option that do not force multi-line signature.

One first step would be to propose such a behavior. However, we MAY not apply to the existing `ensure_fully_multiline` existing option because it might be considered as a breaking change.

We could propose a `force_fully_multiline` option that would be the reverse of `ensure_single_line`. Then, it will be the naming that would be inconsistent. However, we can discuss about it and play with deprecation to solve that.

I'll start working on this to make a proposal.

Note: It does not fully solve this issue considering the debate about the args number/line lenght limit option (https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3168#issuecomment-338914134), but it can be considered later, or not. :stuck_out_tongue:

**Comment by soullivaneuh** [comment]
About the `on_multiline` itself, its name and description would not be very in sync with my latest proposal:

> `on_multiline`: 'Defines how to handle function arguments lists that contain newlines.'

It says the option is for the **current presence** of newlined arguments.

I would propose something like this:

- Option name: `multiline`
- Description: Defines how to handle function arguments lists newlines separation.

Or, we remove the logic from `MethodArgumentSpaceFixer` to create a dedicated fixer. And then, we also may consider https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3168#issuecomment-338914134.

What do you think?

**Comment by soullivaneuh** [comment]
By digging on the `MethodArgumentSpaceFixer` class, I figured out it does not apply only to function declarations but also function calls, which will not fit anybody needs.

Also, the multi-line conversion can also apply the array.

Then I remember why I made this proposal: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3168#issuecomment-338914134

However, the `element` option is restrictive in my opinion because we will not be able to apply different configuration set regarding each element kind.

We may be able to solve the issue by two ways:

1. Transform `array_element`, `func_signature_argument` and `func_call_argument` to option names containing each the `min_items`, `max_length` and `inclusive` options.
2. Create separated fixer for each. E.g. `ArrayMultilineFixer`, `FunctionSignatureMultilineFixer` and `FunctionCallArgumentMultilineFixer`.

I personally prefer the first option as the logic will be seamlessly the same.

This fixer should also be usable for single-line enforcement. For example, if we set `min_items => 2`, the fixer can transform a multi-line signature to single-line if the number of items is 1.
On that case, we may consider the depreciation of `method_argument_space:on_multiline` option completly.

What do you think?

I'll start to work on that soon.

**Comment by soullivaneuh** [comment]
The first option is not acceptable because the `FixerConfigurationResolver` service only works with first level option value.

In another terms, is is not possible to have the `min_items` and `max_length` option being defined as "sub-configuration" for `array_element`, `func_signature_argument` and `func_call_argument`.

We might go back to proposal of https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3168#issuecomment-338914134, but the configuration will be less flexible.

We can also plan to option 2, meaning one fixer for each.

Any feedback would be appreciated. :)

**Comment by jlecordier** [comment]
I have already argued that I am in favor of multiple fixers [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6930#issuecomment-2132839887)

@Wirone argued against it

I'll let you check the discussion on the other [issue](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6930)

**Comment by soullivaneuh** [comment]
@jlecordier Thanks for the feedback!

The PR #6930 concerns only array, and I partially agree with your statement: 

> I think it would be easier to develop fixers independently, and then refactor later once each one is reviewed, so we can more easily identify duplicated code

Having a fixer for each element kind is better for flexibility (e.g. some may want to apply multi-lines on array but not on function calls) with a clear documentation of the option.

Concerning the refactor part, I'd prefer to manage it on the same PR, proposing three fixer at the same time.

**Comment by soullivaneuh** [comment]
> Hey everybody, this fixer is available at https://github.com/Symplify/CodingStandard#arguments-should-be-on-the-samestandalone-line-to-fit-line-length (Symplify\CodingStandard\Fixer\LineLength\BreakMethodArgumentsFixer)
> 
> cc @TomasVotruba

It seems the mentioned fixer does not exist anymore. Instead, I found the [`LineLengthFixer`](https://github.com/symplify/coding-standard/blob/5de526650985cce3c27c9934461df79ef5c7fd16/src/Fixer/LineLength/LineLengthFixer.php).

It apply line breaking on any breakable element (functions, calls, array, use...) regarding a line length rule only.

However, I still think this issue is actual to propose something improved with two main goal:

- Add an option regarding the number of items instead or withing the line length
- Separate fixer for each element to ensure a full configuration flexibility

**Comment by soullivaneuh** [comment]
According to the [fixer definition of simplify](https://github.com/symplify/coding-standard/blob/5de526650985cce3c27c9934461df79ef5c7fd16/src/Fixer/LineLength/LineLengthFixer.php), this kind of fixer can apply on the following elements:

- array (`[]`, `array()`)
- attributes (`#[`)
- function declaration
- function call
- new (`new MyClass()`)
- use (`function foo (...) use (...)`)

For the last 3 mentioned elements, I am hesitating to provide a single fixer for all named `BreakCallLineFixer`. Any feedback is appreciated.
